service/src/main/java/bio/terra/workspace/service/resource/controlled/cloud/aws/sagemakerNotebook/CreateAwsSagemakerNotebookStep.java

```diff
@@ -0,0 +1,85 @@
+package bio.terra.workspace.service.resource.controlled.cloud.aws.sagemakerNotebook;
+
+import bio.terra.common.iam.SamUser;
+import bio.terra.stairway.FlightContext;
+import bio.terra.stairway.FlightMap;
+import bio.terra.stairway.Step;
+import bio.terra.stairway.StepResult;
+import bio.terra.stairway.StepStatus;
+import bio.terra.stairway.exception.RetryException;
+import bio.terra.workspace.common.utils.AwsUtils;
+import bio.terra.workspace.generated.model.ApiAwsSagemakerNotebookCreationParameters;
+import bio.terra.workspace.service.iam.AuthenticatedUserRequest;
+import bio.terra.workspace.service.iam.SamService;
+import bio.terra.workspace.service.workspace.AwsCloudContextService;
+import bio.terra.workspace.service.workspace.flight.WorkspaceFlightMapKeys.ControlledResourceKeys;
+import bio.terra.workspace.service.workspace.model.AwsCloudContext;
+import java.util.Collection;
+import java.util.HashSet;
+import software.amazon.awssdk.arns.Arn;
+import software.amazon.awssdk.auth.credentials.AwsCredentialsProvider;
+import software.amazon.awssdk.services.sts.model.Tag;
+
+public class CreateAwsSagemakerNotebookStep implements Step {
+
+  private final ControlledAwsSagemakerNotebookResource resource;
+  private final AwsCloudContextService awsCloudContextService;
+  private final AuthenticatedUserRequest userRequest;
+  private final SamService samService;
+
+  public CreateAwsSagemakerNotebookStep(
+      ControlledAwsSagemakerNotebookResource resource,
+      AwsCloudContextService awsCloudContextService,
+      AuthenticatedUserRequest userRequest,
+      SamService samService) {
+    this.resource = resource;
+    this.awsCloudContextService = awsCloudContextService;
+    this.userRequest = userRequest;
+    this.samService = samService;
+  }
+
+  @Override
+  public StepResult doStep(FlightContext flightContext)
+      throws InterruptedException, RetryException {
+    AwsCloudContext cloudContext =
+        awsCloudContextService.getRequiredAwsCloudContext(resource.getWorkspaceId());
+
+    AwsCredentialsProvider credentialsProvider =
+        AwsUtils.createWsmCredentialProvider(
+            awsCloudContextService.getRequiredAuthentication(),
+            awsCloudContextService.discoverEnvironment());
+
+    SamUser samUser = samService.getSamUser(userRequest);
+    Collection<Tag> tags = new HashSet<>();
+    AwsUtils.appendUserTags(tags, samUser);
+    AwsUtils.appendResourceTags(tags, cloudContext);
```

Comment:
If it's not too much trouble, could you add the Workspace ID and the CLI Server name as part of `appendResourceTags`?  The startup script will need them to resolve the instance to the WS to bootstrap Terra CLI setup, and expects them with keys `WorkspaceId` and `CliServerName` respectively (see [Infrastructure PR](https://github.com/verily-src/mc-terra-aws-wsm-landingzone/pull/25) for more detail.

Also LMK if you'd prefer me to add this as a follow-on.

Ref:

Reply:
Added both. 
workspaceId (user facing id) - It is available in `resource`, and added to both S3 folders and notebooks (may be useful for S3 folder as well)
CliServerName - likely only needed by notebook, hence added outside the util function, in the step.do()